src/ucs/datastruct/ptr_array.h

```diff
@@ -121,12 +135,125 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
                     !__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index])))
 
 
+static inline int
+__ucs_ptr_array_is_free_and_get_step_size(ucs_ptr_array_elem_t _elem, uint32_t *size_free)
+{
+    int is_free = ((uintptr_t)(_elem) & UCS_PTR_ARRAY_FLAG_FREE);
+
+    if (is_free) {
+       *size_free = _elem >> UCS_PTR_ARRAY_SIZE_FREE_SHIFT;
+    } else {
+       *size_free = 1;
+    }
+
+    return is_free;
+}
+
 /**
  * Iterate over all valid elements in the array.
  */
 #define ucs_ptr_array_for_each(_var, _index, _ptr_array) \
-    for (_index = 0; _index < (_ptr_array)->size; ++_index) \
-         if (!__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index]))) \
+    uint32_t step_size = 0;\
+    for (_index = 0; (_index < (_ptr_array)->size); _index += step_size)\
+       if ((ucs_likely(!__ucs_ptr_array_is_free_and_get_step_size((ucs_ptr_array_elem_t)(_var = (void *)((_ptr_array)->start[_index])), &step_size))))\
 
 
+/*
+   Locked interface
+*/
+
+/* Locked ptr array */
+typedef struct ucs_ptr_array_locked {
```

Comment:
```
typedef struct ucs_ptr_array_locked {
    ucs_ptr_array_t             super;
    ucs_recursive_spinlock_t    lock;
} ucs_ptr_array_locked_t;

Reply:
?

Reply:
Well, I'm not sure what you mean here, please review the commit I'm sending in a few minutes :)